### PR TITLE
devicetree: add DT_NODE_FULL_NAME()

### PR DESCRIPTION
**Pull request by galak**
This macro returns a node's name, given its node identifier.

The node name is useful information for the user to utilize for debug
information, similar to DT_NODE_PATH, or DT_LABEL.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

### CODE REVIEW

File: scripts/dts/gen_defines.py [hunk 1]
```diff
@@ -98,6 +98,9 @@ def main():
             out_comment(f"Node's full path:")
             out_dt_define(f"{node.z_path_id}_PATH", f'"{escape(node.path)}"')
 
+            out_comment(f"Node's name:")
```
Comment (gmarull) [inline scripts/dts/gen_defines.py:101]:
```suggestion
            out_comment("Node's name:")
```

Comment (galak) [reply scripts/dts/gen_defines.py:101]: @gmarull thanks.  I stole the pattern from a few lines above.

**Review by gmarull — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Review by galak — COMMENTED** [review]
Should we call this `DT_NODE_FULLNAME`, `DT_NODE_FQN`?

**Review by pabigot — COMMENTED** [review]
Copied from #32627 since I hadn't seen this:

As I noted in dev-review my main concern is with precise use of devicetree terminology:

> Each node in the devicetree is named according to the following convention:
>
>     node-name@unit-address
>  The node-name component specifies the name of the node....The unit-address component of the name is specific to the bus type on which the node sits.

I'd like to see the combined identifier called something like `DT_NODE_NAME_AT_UNIT_ADDRESS`.   I really would expect `DT_NODE_NAME` to just be the node-name component.

There is a risk this will not be unique, but if it isn't we can make it the SOC maintainer's responsibility to add a `label` property which will take priority.

However if nobody shares my concern, let it be.

**Comment by galak** [comment]
> I'd like to see the combined identifier called something like `DT_NODE_NAME_AT_UNIT_ADDRESS`. I really would expect `DT_NODE_NAME` to just be the node-name component.
> 
> There is a risk this will not be unique, but if it isn't we can make it the SOC maintainer's responsibility to add a `label` property which will take priority.
> 
> However if nobody shares my concern, let it be.

I'm in agreement that the macro should be renamed.  List of possible names:

* `DT_NODE_NAME_AT_UNIT_ADDRESS`
* `DT_NODE_NAME_WITH_UNIT_ADDRESS`
* `DT_NODE_FULLNAME`
* `DT_NODE_FQN`

**Comment by pabigot** [comment]
My preference order by weight:

* `DT_NODE_NAME_WITH_UNIT_ADDRESS` 90%
* `DT_NODE_NAME_AT_UNIT_ADDRESS` 8%
* `DT_NODE_FULLNAME` 2%
* `DT_NODE_FQN` nak

**Comment by erwango** [comment]
I'm ,not sure what the `UNIT` is referring to. What about `DT_NODE_NAME_WITH_SOC_ADDRESS` ?

**Comment by pabigot** [comment]
> I'm ,not sure what the `UNIT` is referring to. What about `DT_NODE_NAME_WITH_SOC_ADDRESS` ?

Nak.  It's nothing necessarily to do with the SoC; the unit address is the first cell of the node's `reg` property (if it has a `reg` property, which is optional).  For some peripheral's it's an address in the SOC address space.  For I2C it's a bus address; for SPI it's a chip-select index.

**Comment by galak** [comment]
> Nak. It's nothing necessarily to do with the SoC; the unit address is the first cell of the node's `reg` property (if it has a `reg` property, which is optional). For some peripherals it's an address in the SOC address space. For I2C it's a bus address; for SPI it's a chip-select index.

As @pabigot this unit-address comes from devicetree specification.

node-name

**Comment by mbolivar-nordic** [comment]
My two cents in order of preference:

- DT_NODE_FULL_NAME
- DT_NODE_NAME_WITH_UNIT_ADDRESS
- DT_NODE_FULLNAME

**Comment by jfischer-no** [comment]
* DT_NODE_FQN
* DT_NODE_FULL_NAME
* DT_NODE_FULLNAME

**Comment by erwango** [comment]
`DT_NODE_FULL_NAME` would be nice. Then people can check definition to see what 'full' refers to.

**Comment by galak** [comment]
Ok, went with `DT_NODE_FULL_NAME`

**Review by jfischer-no — APPROVED** [review]

File: tests/lib/devicetree/api/src/main.c [hunk 2]
```diff
@@ -1625,6 +1625,13 @@ static void test_path(void)
 			     "/test/gpio@deadbeef"), "");
 }
 
+static void test_node_name(void)
+{
+	zassert_true(!strcmp(DT_NODE_FULL_NAME(DT_ROOT), "/"), "");
```
Comment (pabigot) [inline tests/lib/devicetree/api/src/main.c:1630]: Please add this or something similar, so we verify it works when there is no unit address:
```
	zassert_true(!strcmp(DT_NODE_FULL_NAME(TEST_TEMP),
			     "temperature-sensor"), "");
```

**Review by pabigot — CHANGES_REQUESTED** [review]
One test improvement please, otherwise this seems fine.

**Review by mbolivar-nordic — APPROVED** [review]

File: tests/lib/devicetree/api/src/main.c (hunk 2 above)
Comment (galak) [reply tests/lib/devicetree/api/src/main.c:1630]: Added

File: tests/lib/devicetree/api/src/main.c [hunk 3]
```diff
@@ -1625,6 +1625,17 @@ static void test_path(void)
...
+	zassert_true(!strcmp(DT_NODE_FULL_NAME(TEST_DEADBEEF),
+			     "gpio@deadbeef"), "");
+	zassert_true(!strcmp(DT_NODE_FULL_NAME(TEST_TEMP),
+			     "temperature-sensor"), "");
+	zassert_true(strcmp(DT_NODE_FULL_NAME(TEST_REG),
+			     "reg-holder"), "");
```
Comment (pabigot) [inline tests/lib/devicetree/api/src/main.c:1636]: Why isn't this `"reg-holder@9999aaaa"`?

Comment (galak) [reply tests/lib/devicetree/api/src/main.c:1636]: Because I'm testing that it does NOT match.

Comment (pabigot) [reply tests/lib/devicetree/api/src/main.c:1636]: Ah.  A little subtle.

**Review by pabigot — APPROVED** [review]

**Pull request merged**